File: enode/src/main/java/org/enodeframework/messaging/impl/QueuedHandler.java
package org.enodeframework.messaging.impl;

import org.enodeframework.common.function.Action2;
import org.enodeframework.infrastructure.IObjectProxy;

import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

public class QueuedHandler<T extends IObjectProxy> {
    private final Action2<QueuedHandler<T>, T> dispatchToNextHandler;
    private final ConcurrentLinkedQueue<T> handlerQueue;

    public QueuedHandler(List<T> handlers, Action2<QueuedHandler<T>, T> dispatchToNextHandler) {
        handlerQueue = new ConcurrentLinkedQueue<>();
        handlerQueue.addAll(handlers);
        this.dispatchToNextHandler = dispatchToNextHandler;
    }

    public T dequeueHandler() {
        return handlerQueue.poll();
    }

    public void onHandlerFinished(T handler) {
        T nextHandler = dequeueHandler();
        if (nextHandler != null) {
            dispatchToNextHandler.apply(this, nextHandler);
        }
    }
}

File: enode/src/main/java/org/enodeframework/messaging/impl/QueuedHandler.kt
package org.enodeframework.messaging.impl

import org.enodeframework.common.function.Action2
import org.enodeframework.infrastructure.ObjectProxy
import java.util.concurrent.ConcurrentLinkedQueue

class QueuedHandler<T : ObjectProxy?>(handlers: List<T>, dispatchToNextHandler: Action2<QueuedHandler<T>, T>) {
    private val dispatchToNextHandler: Action2<QueuedHandler<T>, T>
    private val handlerQueue: ConcurrentLinkedQueue<T> = ConcurrentLinkedQueue()

    init {
        handlerQueue.addAll(handlers)
        this.dispatchToNextHandler = dispatchToNextHandler
    }

    fun dequeueHandler(): T {
        return handlerQueue.poll()
    }

    fun onHandlerFinished(handler: T) {
        val nextHandler = dequeueHandler()
        if (nextHandler != null) {
            dispatchToNextHandler.apply(this, nextHandler)
        }
    }
}
